Popover: refactor to TypeScript (#43823)
WordPress/gutenberg: packages/components/src/popover/README.md

An object extending a `DOMRect` with an additional optional `ownerDocument` property, used to specify a fixed popover position.

-   Required: No

### `anchorRef`: `Element | PopoverAnchorRefReference | PopoverAnchorRefTopBottom | Range`

Used to specify a fixed popover position. It can be an `Element`, a React reference to an `element`, an object with a `top` and a `bottom` properties (both pointing to elements), or a `range`.

-   Required: No

### `animate`: `boolean`

Whether the popover should animate when opening.

-   Required: No
-   Default: `true`

### `children`: `ReactNode`

The `children` elements rendered as the popover's content.

-   Required: Yes

### `expandOnMobile`: `boolean`

Show the popover fullscreen on mobile viewports.

-   Required: No

### `flip`: `boolean`

Specifies whether the popover should flip across its axis if there isn't space for it in the normal placement.

When the using a 'top' placement, the popover will switch to a 'bottom' placement. When using a 'left' placement, the popover will switch to a `right' placement.

The popover will retain its alignment of 'start' or 'end' when flipping.

-   Required: No
-   Default: `true`

> ### `focusOnMount`: `'firstElement' | boolean`

Author: Removed `container` as a possible option as it looks like it was removed in https://github.com/WordPress/gutenberg/pull/27699

The component accepts the following props. Props not included in this set will be applied to the element wrapping Popover content.

### `__unstableSlotName`: `string`

The name of the Slot in which the popover should be rendered. It should be also passed to the corresponding `PopoverSlot` component.

-   Required: No
-   Default: `"Popover"`

> ### `__unstableShift`: `boolean`

Reviewer: This prop has been flagged as deprecated in `types.ts` should we note that here as well?

Author: Good point. I decided to remove all `__unstable` props from the README, and added the `range` prop instead https://github.com/WordPress/gutenberg/pull/43823/commits/57692b878ae48de0ecc6133fadeb6e02faf321bd

Enables the `Popover` to shift in order to stay in view when meeting the viewport edges.

-   Required: No
-   Default: `false`

### `anchorRect`: `DomRectWithOwnerDocument`

An object extending a `DOMRect` with an additional optional `ownerDocument` property, used to specify a fixed popover position.

-   Required: No

> ### `anchorRef`: `Element | Range | `"firstElement"`PopoverAnchorRefReference | PopoverAnchorRefTopBottom`

Reviewer: ```suggestion
### `anchorRef`: `Element | PopoverAnchorRefReference | PopoverAnchorRefTopBottom | Range`
```

Should this match `types.ts`?

Author: https://github.com/WordPress/gutenberg/pull/43823/commits/8af91037d9b9290cb75e2908bb1e91d25a327dc3

Used to specify a fixed popover position. It can be an `Element`, a React reference to an `element`, an object with a `top` and a `bottom` properties (both pointing to elements), or a `range`.

-   Required: No

### `animate`: `boolean`

Whether the popover should animate when opening.

-   Required: No
-   Default: `true`

### `children`: `ReactNode`

The `children` elements rendered as the popover's content.

> -   Required: No

Reviewer: The type for `children` in `PopoverProps` doesn't explicitly state it is optional. Should this be `Yes`?

Author: https://github.com/WordPress/gutenberg/pull/43823/commits/828cb7469d4a0b7f07d0e301d72b65fee5e49ad1

### `expandOnMobile`: `boolean`

Show the popover fullscreen on mobile viewports.

-   Required: No

### `flip`: `boolean`

Specifies whether the popover should flip across its axis if there isn't space for it in the normal placement.

When the using a 'top' placement, the popover will switch to a 'bottom' placement. When using a 'left' placement, the popover will switch to a `right' placement.

The popover will retain its alignment of 'start' or 'end' when flipping.

-   Required: No
-   Default: `true`

### `focusOnMount`: `'firstElement' | boolean`

> y default, the _first tabblable element_ in the popover will receive focus when it mounts. This is the same as setting this prop to `"firstElement"`.

Reviewer: ```suggestion
By default, the _first tabbable element_ in the popover will receive focus when it mounts. This is the same as setting this prop to `"firstElement"`.
```

Typos.

Author: https://github.com/WordPress/gutenberg/pull/43823/commits/5ef6066306805d191df3a0bdbc7b46d594e8cbc3

Specifying a `false` value disables the focus handling entirely (this should only be done when an appropriately accessible substitute behavior exists).

-   Required: No
-   Default: `"firstElement"`

> ### `onFocusOutside`: `( event: SyntheticEvent ) =>

Reviewer: ```suggestion
### `onFocusOutside`: `( event: SyntheticEvent ) => void`
```
Should we match `types.ts` here and explicitly add `void`?

Author: https://github.com/WordPress/gutenberg/pull/43823/commits/f9f1f5d49e076fe508c77184885d19512c5e9e4d

A callback invoked when the focus leaves the opened popover. This should only be provided in advanced use-cases when a popover should close under specific circumstances (for example, if the new `document.activeElement` is content of or otherwise controlling popover visibility).

When not provided, the `onClose` callback will be called instead.

-   Required: No

> ### `getAnchorRect`: `( fallbackReferenceElement:

Reviewer: ```suggestion
### `getAnchorRect`: `( fallbackReferenceElement: Element | null ) => DomRectWithOwnerDocument
```

Should this match `types.ts`?

Author: https://github.com/WordPress/gutenberg/pull/43823/commits/1d9b8a2a4dff16c698404a3e2891ad969ba4f6fb

A function returning the same value as the one expected by the `anchorRect` prop, used to specify a dynamic popover position.

-   Required: No

### `headerTitle`: `string`

Used to customize the header text shown when the popover is toggled to fullscreen on mobile viewports (see the `expandOnMobile` prop).

-   Required: No

### `isAlternate`: `boolean`

Used to enable a different visual style for the popover.

-   Required: No

> -   Default: `"firstElement"`

Reviewer: ```suggestion
```

There's no `@default` for `isAlternate` in `types.ts`. This also doesn't appear to match the `boolean` type. Should this be here?

Author: https://github.com/WordPress/gutenberg/pull/43823/commits/67d8fe7db0dff0126fdab19912b1299dc2213ffb
